hazelcast/src/main/java/com/hazelcast/spi/tenantcontrol/TenantControl.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.tenantcontrol;
+
+import com.hazelcast.spi.annotation.Beta;
+import com.hazelcast.spi.impl.NoopTenantControl;
+
+import java.io.Closeable;
+import java.io.Serializable;
+
+/**
+ * Hooks for multi-tenancy for application servers
+ * Hazelcast uses separate threads to invoke operations
+ * this interface acts a hook to establish a thread-local tenant context
+ * so that operation invocations into application servers are handled correctly
+ * This is used by application servers to establish thread context for class loading,
+ * CDI, EJB and JPA invocations
+ *
+ * @author lprimak
+ */
+@Beta
+public interface TenantControl extends Serializable {
```

Comment:
It seems hard to see if this `TenantControl` api is mature enough without a concrete implementation, maybe it is because i'm not familiar with this tenant mechanism. 
In the current form of api, for example, `saveCurrentTenant` is saving current tenant and returning new one, at the same time it wants to know how to destroy objects. This is complicated to me. It is doing lots of things.

Reply:
What I am trying to do with this is almost 2 years in the making now
At this point anything is better than nothing